### ISSUE TITLE
Pool object mounted without "password" property

### ISSUE THREAD

**Issue by aruan-cuenca-emasoftware** [brianc/node-postgres#2488]
**Node**: v15.11
**Postgres**: v10.6

**TL;DR:** Do the Pool objects receive the `password` configuration when getting mounted in all conditions? Or is there a scenario where the property is omitted?

Hi, I'm building a backend in TypeScript with [NestJS](https://github.com/nestjs/nest) using their default caching technique based on [cache-manager](https://github.com/BryanDonovan/node-cache-manager) to store a Pool instance, like this:
```javascript
const configs = {
  database: 'db',
  host: 'localhost',
  password: 'password',
  port: 5432,
  user: 'postgres',
}
const pool = new pg.Pool(configs);
await this.cacheManager.set('postgre_pool', pool);
```
The `pool` object contains the `options` attribute with all configurations.
```javascript
options {
  database: 'db',
  host: 'localhost',
  user: 'postgres',
  password: 'password',
  port: 5432
}
```
But when I retrieve it from the memory, **all configurations are still present, expect `password`**:
```javascript
await this.cacheManager.get<pg.Pool>('postgre_pool');
```
results in:
```javascript
options {
  database: 'db',
  host: 'localhost',
  user: 'postgres',
  port: 5432
}
```
Then, when I try to get a client connection with `const client = await pool.connect()`, I get this error:
```
error: empty password returned by client
    at Parser.parseErrorMessage (C:\temp\pex-js\node_modules\pg-protocol\dist\parser.js:278:15)
    at Parser.handlePacket (C:\temp\pex-js\node_modules\pg-protocol\dist\parser.js:126:29)
    at Parser.parse (C:\temp\pex-js\node_modules\pg-protocol\dist\parser.js:39:38)
    at Socket.<anonymous> (C:\temp\pex-js\node_modules\pg-protocol\dist\index.js:10:42)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:271:9)
    at Socket.Readable.push (_stream_readable.js:212:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
    at TCP.callbackTrampoline (internal/async_hooks.js:120:14)
```
So, I thought it could be related to the caching package. In order to test this, I created a structure similar to the Pool class, like this:
```javascript
interface BaseOptions {
  password?: string | (() => Promise<string> | string);
}
interface OptionsTeste extends BaseOptions {
  user: string;
}
class Teste {
  constructor(options?: OptionsTeste) {
    this.options = options;
  }
  options?: OptionsTeste;
}

const testObject = new Teste({ user: 'foo', password: 'bar' });
await this.cacheManager.set('test_object', testObject);
const cachedObject = await this.cacheManager.get<Teste>('test_object');
```
And the resulting object contains all the expected properties. So is there a problem with the way the Pool objects are created? Because all other properties are present in the `pool` object. Or is it really a problem with the caching?

Thank you!

**Comment by charmander** [comment]
A glance at cache-manager suggests that it’s for caching data and that it’d be inappropriate to store a `pg.Pool` inside it. Why do you need to do that? (Yes, the cache is breaking it.)

**Comment by aruan-cuenca-emasoftware** [comment]
I'm storing connection pools in order to make connections to the database faster. The user has multiple db configuration/credentials stored on a database (for mssql, mysql, odbc, oracle, postgre, etc.); when the request arrives, the service layer is responsible for getting the credentials, creating the connections and managing them.
To avoid creating and destroying those connections every request, I decided to use cache-manager to keep the created pools, so I just need to retrieve them from cache and get a connection.
It works well for all pool objects and even for this one, as the returned `pg.Pool` object still has **all** properties, except for password (that's why I thought it wouldn't be a caching problem).

**Comment by boromisp** [comment]
Have you tried to disable cloning the cached value in the [cache-manager](https://github.com/BryanDonovan/node-cache-manager)?
https://github.com/BryanDonovan/node-cache-manager/blob/f4991cec7fb6be82dd6a98c66423e4a05f1bed03/lib/stores/memory.js#L55

**Comment by brianc** [comment]
The password is non-enumerable in the config to prevent it being accidentally logged out.  Does the cache manager actually serialize objects into itself? Because the pool is non-serializable.  Typically one handles it by keeping a singleton to the pool (or a factory function that is memoized & returns a pool per database or user or whatever) - if the cache manager is in memory that's fine, but if it's serializing to something that's likely why you're losing the password property.

**Comment by charmander** [comment]
What @brianc said! You can store the pool *configuration* in the cache-manager if you want and pass that to `new Pool`, but the pool object itself needs to be stored as a JavaScript value rather than a cache that serializes/deserializes/tries to share itself among multiple Node processes.

**Comment by aruan-cuenca-emasoftware** [comment]
@boromisp NestJS hides this layer. I can only choose the store I want to use. The only way I could change this option, is using `CacheStoreFactory.create()` to create a `CacheStore` passing arguments that aren't clear:
```javascript
export interface CacheStoreFactory {
    /**
     * Return a configured cache store.
     *
     * @param args Cache manager options received from `CacheModule.register()`
     * or `CacheModule.registerAcync()`
     */
    create(args: LiteralObject): CacheStore;
}
export interface LiteralObject {
    [key: string]: any;
}
```
The docs say nothing about it, and creating one with no documentation would only lead me to more problems.

@brianc The storage is made in memory, as they mention [in node-cache-manager](https://github.com/BryanDonovan/node-cache-manager#overview)
> **Second**, node-cache-manager features a built-in memory cache (using [node-lru-cache](https://github.com/isaacs/node-lru-cache))

And [in NestJS docs](https://docs.nestjs.com/techniques/caching#in-memory-cache)
>Nest provides a unified API for various cache storage providers. The built-in one is an in-memory data store.

But as you said, the property is non-enumerable, so I guess that's the reason it's not getting retrieved.

**Comment by aruan-cuenca-emasoftware** [comment]
@charmander You're right. I already cache the configs, so it's already a performance gain. I'll find a better way to use the `Pool` other than caching. 

Thank you, guys!
